adapters/silverpush/silverpush.go

```diff
@@ -0,0 +1,344 @@
+
+	if impExt.BidFloor == 0 {
+		if imp.Banner != nil {
+			imp.BidFloor = 0.05
+		} else if imp.Video != nil {
+			imp.BidFloor = 0.1
+		}
+	} else {
+		imp.BidFloor = impExt.BidFloor
+	}
+
+	if imp.Banner != nil {
+		bannerCopy, err := setBannerDimension(imp.Banner)
+		if err != nil {
+			return err
+		}
+		imp.Banner = bannerCopy
+	}
+
+	if imp.Video != nil {
+		videoCopy, err := checkVideoDimension(imp.Video)
+		if err != nil {
+			return err
+		}
+		imp.Video = videoCopy
+	}
+
+	return nil
+}
+
+func checkVideoDimension(video *openrtb2.Video) (*openrtb2.Video, error) {
+	videoCopy := *video
+	if videoCopy.MaxDuration == 0 {
+		videoCopy.MaxDuration = 120
+	}
+	if videoCopy.MaxDuration < videoCopy.MinDuration {
+		videoCopy.MaxDuration = videoCopy.MinDuration
+		videoCopy.MinDuration = 0
+	}
+	if videoCopy.API == nil || videoCopy.MIMEs == nil || videoCopy.Protocols == nil || videoCopy.MinDuration < 0 {
+		return nil, &errortypes.BadInput{Message: "Invalid or missing video field(s)"}
+	}
+	return &videoCopy, nil
+}
+
+func setBannerDimension(banner *openrtb2.Banner) (*openrtb2.Banner, error) {
+	if banner.W != nil && banner.H != nil {
+		return banner, nil
+	}
+	if len(banner.Format) == 0 {
+		return banner, &errortypes.BadInput{Message: "No sizes provided for Banner."}
+	}
+	bannerCopy := *banner
+	bannerCopy.W = openrtb2.Int64Ptr(banner.Format[0].W)
+	bannerCopy.H = openrtb2.Int64Ptr(banner.Format[0].H)
+
+	return &bannerCopy, nil
+}
+
+func setPublisherId(req *openrtb2.BidRequest, imp *openrtb2.Imp, impExt *openrtb_ext.ImpExtSilverpush) error {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	if err := json.Unmarshal(bidderExt.Bidder, impExt); err != nil {
+		return &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	if impExt.PublisherId == "" {
+		return &errortypes.BadInput{Message: "Missing publisherId parameter."}
+	}
+	if req.Site != nil {
+		siteCopy := *req.Site
+		if siteCopy.Publisher == nil {
+			siteCopy.Publisher = &openrtb2.Publisher{ID: impExt.PublisherId}
+		} else {
+			publisher := *siteCopy.Publisher
+			publisher.ID = impExt.PublisherId
+			siteCopy.Publisher = &publisher
+		}
+		req.Site = &siteCopy
+
+	} else if req.App != nil {
+		appCopy := *req.App
+		if appCopy.Publisher == nil {
+			appCopy.Publisher = &openrtb2.Publisher{ID: impExt.PublisherId}
+		} else {
+			publisher := *appCopy.Publisher
+			publisher.ID = impExt.PublisherId
+			appCopy.Publisher = &publisher
+		}
+		appCopy.Publisher = &openrtb2.Publisher{ID: impExt.PublisherId}
+		req.App = &appCopy
+
+	}
+
+	return nil
+}
+
+func impressionByMediaType(imp *openrtb2.Imp) ([]openrtb2.Imp, error) {
+
+	if imp.Banner == nil && imp.Video == nil {
+		return nil, &errortypes.BadInput{Message: "Invalid MediaType. SilverPush only supports Banner, Video"}
+	}
+	imps := make([]openrtb2.Imp, 0, 2)
```

Comment:
I see that you are initialising `imps` to `2` imps however your function will always return just single impression based on your checks below. I think you might wanna change this function definiton as below - 
```

func getImpressionByMediaType(imp *openrtb2.Imp) (openrtb2.Imp) {
              impCopy := *imp
              if imp.Banner != nil {
		impCopy.Video = nil
	     }
             if imp.Video != nil {
		impCopy.Banner = nil
	     }
	return impCopy
```

Reply:
Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)